README.md
# FRC Team 744 Survey PWA
## (currently GHP demo)
A progressive web application (PWA) made to streamline both survey times used during competitions

## Install Instructions
### iOS -- Safari (iPhone and iPad)
1. Click the Share Menu
2. Click "Add to Home Screen"
3. Leave app name as written
4. Click "Add"

### Android
TBD

The PWA is now available on your device, even offline!

## More to Come!
- landing page
- match survey (to be worked on during build season proper)
- proper styling lol (currently is in Proof of Concept/Install Testing stage)

## Changelog
For changelog, click [here](https://github.com/arifire21/744-survey-pwa/blob/main/changelog.md).


changelog.md
# Pre-Release
## v0.1
- only contains basics of pit survey form (no real styling)
- meant as a proof of concept for PWA functionality
- pushed to GitHub to test GHPages hosting abilities and PWA install process on other devices

## v0.1.1
- GHP "breaks" default SW routing, changed from root "/" to "/744-survey-pwa/" to match
- changed print of past logs to show all values
- fixed timestamp being out of scope of the push method

## v0.1.2
- implementing autocomplete from a static list that will be manually pushed (due to possible unreliable internet)

## v0.2.2
- added homescreen
- added images to drivetrain input, custom styling to make labels look like buttons
- added "other" to drivetrain type, with text input
- made media rule to start adjusting styling for ipad / mobile layout(s)
- changed styling on buttons to let form ones center

## v0.2.3
- changed main font family to Tahoma
- changed button structure to be proper button tags (a's on forms kept clipping)
- added version number to homescreen, changed cache name

sw.js
const VERSION = "v0.2.3";
const CACHE_NAME = `744-survey-${VERSION}`

const APP_STATIC_RESOURCES = [
    "/",
    "/index.html",
    "/styles.css",
    "/app.js",
    "/first_round_list.json",
    "/icon-512x512.png",
];

//install event
self.addEventListener("install", (event) => {
  event.waitUntil(
    (async () => {
      //retrieve, add responses to cache
      const cache = await caches.open(CACHE_NAME);
      cache.addAll(APP_STATIC_RESOURCES);
    })(),
  );
});

//activate event
self.addEventListener("activate", (event) => {
  event.waitUntil(
    (async () => {
      const names = await caches.keys();
      await Promise.all(
        names.map((name) => {
          //check if active, if not, delete
          if (name !== CACHE_NAME) {
            return caches.delete(name);
          }
        }),
      );
      //set sw itself to be new controller
      await clients.claim();
    })(),
  );
});

//fetch event
//do not make requests if already online
self.addEventListener("fetch", (event) => {
  // when seeking an HTML page
  if (event.request.mode === "navigate") {
    // Return to the index.html page
    event.respondWith(caches.match("/"));
    return;
  }

  // For every other request type
  event.respondWith(
    (async () => {
      const cache = await caches.open(CACHE_NAME);
      const cachedResponse = await cache.match(event.request.url);
      if (cachedResponse) {
        // Return the cached response if it's available.
        return cachedResponse;
      }
      // Respond with a HTTP 404 response status.
      return new Response(null, { status: 404 });
    })(),
  );
});


app.js
//import teams from './data.json';

//form id's
const form = document.getElementsByTagName('form')[0];
const pastLogContainer = document.getElementById("past-logs");

// required inputs -- only checkboxes, single radio values grabbed directly in method itself
const teamNumberInput = document.getElementById('team-number');
const autocompleteResults = document.getElementById("auto-results");
const startPositionsInput = document.getElementsByName('start-pos');
const scoreHeightsInput = document.getElementsByName('score-height');
const nameInput = document.getElementById('name');

//optional inputs
const feedbackContainer = document.getElementById('feedback-container');
const feedbackInput = document.getElementById('feedback');

//for the conditional feedback textarea
const investmentYesInput = document.getElementsByName('final-inves')[0];
const investmentNoInput = document.getElementsByName('final-inves')[1];

const firstRoundTeams = ["744", "179", "180", "1", "2", "3", "59", "11111", "1234567890"];

//adjust some stylingggggg
autocompleteResults.style.width = teamNumberInput.style.width + 'px';
// autocompleteResults.offsetLeft = teamNumberInput.offsetLeft;
// autocompleteResults.offsetTop = teamNumberInput.offsetTop;

//apparently this doesnt work on IOS (which is the current intended platform so)
//behavior for handling install stuff
//pre-install prep
// let installPrompt = null;
// const installParagraph = document.getElementById('#install-prompt');
// const installButton = document.getElementById('#install');

// window.addEventListener("beforeinstallprompt", (event) => {
//   //dont let browser display native toolbar ui
//   event.preventDefault();
//   //reference for later
//   installPrompt = event;
//   installButton.removeAttribute("hidden");
//   installParagraph.style.display = "block";
// });

// //actual install trigger
// installButton.addEventListener("click", async () => {
//   if (!installPrompt) {
//     return;
//   }
//   const result = await installPrompt.prompt();
//   console.log(`Install prompt was: ${result.outcome}`);
//   //only once promise allowed per instance, so reset
//   disableInAppInstallPrompt();
// });

//browser might still offer even if installed
window.addEventListener("appinstalled", () => {
  disableInAppInstallPrompt();
});

function disableInAppInstallPrompt() {
  installPrompt = null;
  installButton.setAttribute("hidden", "");
}

//and now back to our regularly scheduled formstuffs
var investment = ''
investmentYesInput.addEventListener('click', function(){
  feedbackContainer.style.display = 'block';
  // investment = investmentYesInput.value;
})

investmentNoInput.addEventListener('click', function(){
  feedbackContainer.style.display = 'none';
  feedbackInput.value = '';
  // investment = investmentNoInput.value;
})

//autocomplete behavior
function getResults(input) {
  const results = [];
  for (i = 0; i < firstRoundTeams.length; i++) {
    if (input === firstRoundTeams[i].slice(0, input.length)) {
      results.push(firstRoundTeams[i]);
    }
  }
  return results;
}

teamNumberInput.addEventListener('input', (event) =>{
// teamNumberInput.oninput = function () {
  console.log(teamNumberInput.value)
  let results = [];
  let userInput = teamNumberInput.value;
  autocompleteResults.innerHTML = "";
  if(userInput.length == 0) {
    autocompleteResults.style.display = "none";
  }
  if (userInput.length > 0) {
    results = getResults(userInput);
    autocompleteResults.style.display = "block";
    if(results.length == 0){
      autocompleteResults.innerHTML = 'No results.';
    }else {
      for (i = 0; i < results.length; i++) {
        autocompleteResults.innerHTML += "<li>" + results[i] + "</li>";
      }
    }
  }
})

autocompleteResults.addEventListener('click', (event) => {
// autocompleteResults.onclick = function (event) {
  const setValue = event.target.innerText;
  teamNumberInput.value = setValue;
  this.innerHTML = "";
  autocompleteResults.style.display = "none";
})

// Listen to form submissions.
form.addEventListener("submit", (event) => {
  // Prevent the form from submitting to the server since everything is client-side.
  event.preventDefault();

  // Get values
  const teamNumber = teamNumberInput.value;
  const drivetrain = document.querySelector('input[name="drivetrain"]:checked').value;
  const vision = document.querySelector('input[name="vision"]:checked').value;
  const climbHeight = document.querySelector('input[name="climb-height"]:checked').value;
  const gamePieces = document.querySelector('input[name="game-pieces"]:checked').value;
  const pickup = document.querySelector('input[name="pickup"]:checked').value;
  const name = nameInput.value;
  const investment = document.querySelector('input[name="final-inves"]:checked').value;
  let startPositions = ''
  let scoreHeights = ''
  let timestamp = ''; //pre-submission

  startPositionsInput.forEach(position => {
    if(position.checked && startPositions === ''){
      startPositions = position.value;
    } else if(position.checked){
      startPositions = startPositions.concat(`, ${position.value}`)
    }
  });
  
  scoreHeightsInput.forEach(height => {
    if(height.checked && scoreHeights === ''){
      scoreHeights = height.value;
    } else if(height.checked){
      scoreHeights = scoreHeights.concat(`, ${height.value}`)
    }
  });


  if (checkRequired(teamNumber, drivetrain, startPositions, vision, scoreHeights, climbHeight, gamePieces, pickup, name, investment)) {
    // If req fields are empty, exit.
    return;
  } else {
    //generate timestamp that will be logged/displayed
    timestamp = new Date(Date.now()).toLocaleString('en-US', {timeZone: 'America/New_York'})
  }

  // Store in our client-side storage.
  storeNewLog(timestamp, teamNumber, drivetrain, vision, scoreHeights, climbHeight, gamePieces, pickup, name, investment);

  // Refresh the UI.
  renderPastLogs();

  // Reset the form.
  form.reset();
});

function checkRequired(teamNumber, drivetrain, startPositions, vision, scoreHeights, climbHeight, gamePieces, pickup, name, investment) {
    if (!teamNumber || !drivetrain || !startPositions || !vision || !climbHeight || !scoreHeights || !gamePieces || !pickup || !name || !investment) {

      // To make the validation robust we could:
      // 1. add error messaging based on error type
      // 2. Alert assistive technology users about the error
      // 3. move focus to the error location
      // instead, for now, we clear the fields if any are missing
      console.log(`#: ${teamNumber}, DT: ${drivetrain}, start: ${startPositions}, vis: ${vision}, CH: ${climbHeight}, SH: ${scoreHeights}, pieces: ${gamePieces}, pickup: ${pickup}, name: ${name}, investment: ${investment}`)
      console.log('invalid elements')
      // form.reset();
      return true;
    }
    // else
    console.log(`#: ${teamNumber}, DT: ${drivetrain}, start: ${startPositions}, vis: ${vision}, CH: ${climbHeight}, SH: ${scoreHeights}, pieces: ${gamePieces}, pickup: ${pickup}, name: ${name}, investment: ${investment}`)
    return false;
}

function checkOptional(){
  //todo. meant for form validation/sanitizing of textarea
}
  
  // Add the storage key as an app-wide constant
const STORAGE_KEY = "robotics-pwa-pit";

function storeNewLog(timestamp, teamNumber, drivetrain, startPositions, vision, climbHeight, scoreHeights, gamePieces, pickup, name, investment) {
  // Get data from storage.
  const logs = getAllStoredLogs();

  // Add the new log object to the end of the array of log objects.
  logs.push({timestamp, teamNumber, drivetrain, startPositions, vision, climbHeight, scoreHeights, gamePieces, pickup, name, investment });

  // Sort the array so that logs are ordered by timestamp for now, may change to team number
  logs.sort((a, b) => {
    return new Date(b.timestamp) - new Date(a.timestamp);
  });

  // Store the updated array back in the storage.
  window.localStorage.setItem(STORAGE_KEY, JSON.stringify(logs));
}

function getAllStoredLogs() {
  // Get the string of log data from localStorage
  const data = window.localStorage.getItem(STORAGE_KEY);

  // If no logs were stored, default to an empty array
  // otherwise, return the stored data as parsed JSON
  const logs = data ? JSON.parse(data) : [];

  return logs;
}

function renderPastLogs() {
    // get the parsed string of logs, or an empty array.
    const logs = getAllStoredLogs();
  
    // exit if there are no logs
    if (logs.length === 0) {
      return;
    }
  
    // Clear the list of past logs, since we're going to re-render it.
    pastLogContainer.innerHTML = "";
  
    const pastLogHeader = document.createElement("h2");
    pastLogHeader.textContent = "Past Logs";
  
    const pastLogList = document.createElement("ul");
  
    // Loop over all logs and render them.
    logs.forEach((log) => {
      const logEl = document.createElement("li");
      logEl.textContent = `[${log.timestamp}] Team: ${log.teamNumber},  DT choice: ${log.drivetrain}, Preferred start position(s): ${log.startPositions}, Vision: ${log.vision}, Highest climb height: ${log.climbHeight}, Scoring height(s): ${log.scoreHeights}, Game piece(s): ${log.gamePieces}, Pickup location(s): ${log.pickup}, name: ${log.name}, Is surveyor invested in ${log.teamNumber}?: ${log.investment}`;
      pastLogList.appendChild(logEl);
    });
  
    pastLogContainer.appendChild(pastLogHeader);
    pastLogContainer.appendChild(pastLogList);
  }
  

pit_survey/pit_code.js
//import teams from './data.json';

//form id's
const form = document.getElementsByTagName('form')[0];
const pastLogContainer = document.getElementById("past-logs");

// required inputs -- only checkboxes, single radio values grabbed directly in method itself
const teamNumberInput = document.getElementById('team-number');
const autocompleteResults = document.getElementById("auto-results");
const startPositionsInput = document.getElementsByName('start-pos');
const scoreHeightsInput = document.getElementsByName('score-height');
const nameInput = document.getElementById('name');

//for listeners
const investmentInput = document.getElementsByName('final-inves');
const drivetrainInput = document.getElementsByName('drivetrain');
const customRadiButtons = document.getElementsByClassName('custom-radio');

//optional inputs
const feedbackContainer = document.getElementById('feedback-container');
const feedbackInput = document.getElementById('feedback');
const otherInput = document.getElementById('other-desc');

const firstRoundTeams = ["744", "179", "180", "1", "2", "3", "59", "11111", "1234567890"];

//attaching all listeners here
investmentInput.forEach(function (element){
  element.addEventListener('click', function(){
    if(element.value === 'Yes'){
      console.log('yes')
      feedbackContainer.style.display = 'block';
    } else if(element.value !='Yes' && feedbackContainer.style.display === 'block') {
      feedbackContainer.style.display = 'none';
    }
  })
})

//attach to custom buttons first

drivetrainInput.forEach(function (element){
  element.addEventListener('click', function(){
    if(element.value === 'Other'){
      console.log('yes')
      otherInput.style.display = 'block';
    } else if(element.value !='Other' && otherInput.style.display === 'block') {
      otherInput.style.display = 'none';
    }
  })
})

//autocomplete behavior
function getResults(input) {
  const results = [];
  for (i = 0; i < firstRoundTeams.length; i++) {
    if (input === firstRoundTeams[i].slice(0, input.length)) {
      results.push(firstRoundTeams[i]);
    }
  }
  return results;
}

teamNumberInput.addEventListener('input', (event) =>{
// teamNumberInput.oninput = function () {
  console.log(teamNumberInput.value)
  let results = [];
  let userInput = teamNumberInput.value;
  autocompleteResults.innerHTML = "";
  if(userInput.length == 0) {
    autocompleteResults.style.display = "none";
  }
  if (userInput.length > 0) {
    results = getResults(userInput);
    autocompleteResults.style.display = "block";
    if(results.length == 0){
      autocompleteResults.innerHTML = 'No results.';
    }else {
      for (i = 0; i < results.length; i++) {
        autocompleteResults.innerHTML += "<li>" + results[i] + "</li>";
      }
    }
  }
})

autocompleteResults.addEventListener('click', (event) => {
// autocompleteResults.onclick = function (event) {
  const setValue = event.target.innerText;
  teamNumberInput.value = setValue;
  this.innerHTML = "";
  autocompleteResults.style.display = "none";
})

// Listen to form submissions.
form.addEventListener("submit", (event) => {
  // Prevent the form from submitting to the server since everything is client-side.
  event.preventDefault();

  // Get values
  const teamNumber = teamNumberInput.value;
  const drivetrain = document.querySelector('input[name="drivetrain"]:checked').value;
  const vision = document.querySelector('input[name="vision"]:checked').value;
  const climbHeight = document.querySelector('input[name="climb-height"]:checked').value;
  const gamePieces = document.querySelector('input[name="game-pieces"]:checked').value;
  const pickup = document.querySelector('input[name="pickup"]:checked').value;
  const name = nameInput.value;
  const investment = document.querySelector('input[name="final-inves"]:checked').value;
  let startPositions = ''
  let scoreHeights = ''
  let timestamp = ''; //pre-submission

  // TODO GET OTHER INPUT

  startPositionsInput.forEach(position => {
    if(position.checked && startPositions === ''){
      startPositions = position.value;
    } else if(position.checked){
      startPositions = startPositions.concat(`, ${position.value}`)
    }
  });
  
  scoreHeightsInput.forEach(height => {
    if(height.checked && scoreHeights === ''){
      scoreHeights = height.value;
    } else if(height.checked){
      scoreHeights = scoreHeights.concat(`, ${height.value}`)
    }
  });


  if (checkRequired(teamNumber, drivetrain, startPositions, vision, scoreHeights, climbHeight, gamePieces, pickup, name, investment)) {
    // If req fields are empty, exit.
    return;
  } else {
    //generate timestamp that will be logged/displayed
    timestamp = new Date(Date.now()).toLocaleString('en-US', {timeZone: 'America/New_York'})
  }

  // Store in our client-side storage.
  storeNewLog(timestamp, teamNumber, drivetrain, vision, scoreHeights, climbHeight, gamePieces, pickup, name, investment);

  // Refresh the UI.
  renderPastLogs();

  // Reset the form.
  form.reset();
});

function checkRequired(teamNumber, drivetrain, startPositions, vision, scoreHeights, climbHeight, gamePieces, pickup, name, investment) {
    if (!teamNumber || !drivetrain || !startPositions || !vision || !climbHeight || !scoreHeights || !gamePieces || !pickup || !name || !investment) {

      // To make the validation robust we could:
      // 1. add error messaging based on error type
      // 2. Alert assistive technology users about the error
      // 3. move focus to the error location
      // instead, for now, we clear the fields if any are missing
      console.log(`#: ${teamNumber}, DT: ${drivetrain}, start: ${startPositions}, vis: ${vision}, CH: ${climbHeight}, SH: ${scoreHeights}, pieces: ${gamePieces}, pickup: ${pickup}, name: ${name}, investment: ${investment}`)
      console.log('invalid elements')
      // form.reset();
      return true;
    }
    // else
    console.log(`#: ${teamNumber}, DT: ${drivetrain}, start: ${startPositions}, vis: ${vision}, CH: ${climbHeight}, SH: ${scoreHeights}, pieces: ${gamePieces}, pickup: ${pickup}, name: ${name}, investment: ${investment}`)
    return false;
}

function checkOptional(){
  //todo. meant for form validation/sanitizing of textarea
}
  
  // Add the storage key as an app-wide constant
const STORAGE_KEY = "robotics-pwa-pit";

function storeNewLog(timestamp, teamNumber, drivetrain, startPositions, vision, climbHeight, scoreHeights, gamePieces, pickup, name, investment) {
  // Get data from storage.
  const logs = getAllStoredLogs();

  // Add the new log object to the end of the array of log objects.
  logs.push({timestamp, teamNumber, drivetrain, startPositions, vision, climbHeight, scoreHeights, gamePieces, pickup, name, investment });

  // Sort the array so that logs are ordered by timestamp for now, may change to team number
  logs.sort((a, b) => {
    return new Date(b.timestamp) - new Date(a.timestamp);
  });

  // Store the updated array back in the storage.
  window.localStorage.setItem(STORAGE_KEY, JSON.stringify(logs));
}

function getAllStoredLogs() {
  // Get the string of log data from localStorage
  const data = window.localStorage.getItem(STORAGE_KEY);

  // If no logs were stored, default to an empty array
  // otherwise, return the stored data as parsed JSON
  const logs = data ? JSON.parse(data) : [];

  return logs;
}

function renderPastLogs() {
    // get the parsed string of logs, or an empty array.
    const logs = getAllStoredLogs();
  
    // exit if there are no logs
    if (logs.length === 0) {
      return;
    }
  
    // Clear the list of past logs, since we're going to re-render it.
    pastLogContainer.innerHTML = "";
  
    const pastLogHeader = document.createElement("h2");
    pastLogHeader.textContent = "Past Logs";
  
    const pastLogList = document.createElement("ul");
  
    // Loop over all logs and render them.
    logs.forEach((log) => {
      const logEl = document.createElement("li");
      logEl.textContent = `[${log.timestamp}] Team: ${log.teamNumber},  DT choice: ${log.drivetrain}, Preferred start position(s): ${log.startPositions}, Vision: ${log.vision}, Highest climb height: ${log.climbHeight}, Scoring height(s): ${log.scoreHeights}, Game piece(s): ${log.gamePieces}, Pickup location(s): ${log.pickup}, name: ${log.name}, Is surveyor invested in ${log.teamNumber}?: ${log.investment}`;
      pastLogList.appendChild(logEl);
    });
  
    pastLogContainer.appendChild(pastLogHeader);
    pastLogContainer.appendChild(pastLogList);
  }
  